polygon/heimdall/service.go

```diff
@@ -0,0 +1,158 @@
+
+	libcommon "github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/erigon-lib/kv"
+	"github.com/ledgerwatch/erigon/polygon/polygoncommon"
+	"github.com/ledgerwatch/erigon/turbo/services"
+)
+
+type Service interface {
+	Heimdall
+	Run(ctx context.Context) error
+}
+
+type service struct {
+	scraper *Scraper
+
+	checkpointStore entityStore
+	milestoneStore  entityStore
+	spanStore       entityStore
+}
+
+func newCheckpointStore(tx kv.RwTx, reader services.BorCheckpointReader, blockNumToIdIndexFactory func(context.Context) (*RangeIndex, error)) entityStore {
+	makeEntity := func() Entity { return new(Checkpoint) }
+	return newEntityStore(tx, kv.BorCheckpoints, makeEntity, reader.LastCheckpointId, reader.Checkpoint, blockNumToIdIndexFactory)
+}
+
+func newMilestoneStore(tx kv.RwTx, reader services.BorMilestoneReader, blockNumToIdIndexFactory func(context.Context) (*RangeIndex, error)) entityStore {
+	makeEntity := func() Entity { return new(Milestone) }
+	return newEntityStore(tx, kv.BorMilestones, makeEntity, reader.LastMilestoneId, reader.Milestone, blockNumToIdIndexFactory)
+}
+
+func newSpanStore(tx kv.RwTx, reader services.BorSpanReader, blockNumToIdIndexFactory func(context.Context) (*RangeIndex, error)) entityStore {
+	makeEntity := func() Entity { return new(Span) }
+	return newEntityStore(tx, kv.BorSpans, makeEntity, reader.LastSpanId, reader.Span, blockNumToIdIndexFactory)
+}
+
+func NewService(
+	heimdallUrl string,
+	tmpDir string,
+	logger log.Logger,
+) Service {
+	// TODO: implementing these is an upcoming task
+	txProvider := func() kv.RwTx { /* TODO */ return nil }
+	readerProvider := func() reader { /* TODO */ return nil }
+
+	tx := txProvider()
+	if tx == nil {
+		// TODO: implement and remove
+		logger.Warn("heimdall.Service txProvider is not implemented yet")
+		return nil
+	}
+	reader := readerProvider()
+	if reader == nil {
+		// TODO: implement and remove
+		logger.Warn("heimdall.Service readerProvider is not implemented yet")
+		return nil
+	}
+
+	blockNumToIdIndexFactory := func(ctx context.Context) (*RangeIndex, error) {
+		return NewRangeIndex(ctx, tmpDir, logger)
+	}
+
+	checkpointStore := newCheckpointStore(tx, reader, blockNumToIdIndexFactory)
+	milestoneStore := newMilestoneStore(tx, reader, blockNumToIdIndexFactory)
+	spanStore := newSpanStore(tx, reader, blockNumToIdIndexFactory)
+
+	client := NewHeimdallClient(heimdallUrl, logger)
+	scraper := NewScraper(
+		checkpointStore,
+		milestoneStore,
+		spanStore,
+		client,
+		1*time.Second,
+		logger,
+	)
+
+	return &service{
+		scraper: scraper,
+
+		checkpointStore: checkpointStore,
+		milestoneStore:  milestoneStore,
+		spanStore:       spanStore,
+	}
+}
+
+func (s *service) FetchLatestSpan(ctx context.Context) (*Span, error) {
+	s.scraper.Synchronize(ctx)
+	entity, err := s.spanStore.GetLastEntity(ctx)
+	return downcastSpanEntity(entity), err
+}
+
+func castEntityToWaypoint(entity Entity) Waypoint {
+	return entity.(Waypoint)
```

Comment:
there is a lot of unsafe casting happening...
would you address this in follow ups please?

Reply:
:+1: